header/p2p/session.go

```diff
@@ -0,0 +1,213 @@
+	headers chan []*header.ExtendedHeader,
+) {
+	// request headers from the remote peer
+	r, size, duration, err := s.requestHeaders(s.ctx, stat.peerID, req)
+	if err != nil {
+		if err == context.Canceled {
+			return
+		}
+		log.Errorw("requesting headers from peer failed."+
+			"Retrying the request from different peer", "pID", stat.peerID, "err", err)
+		s.reqCh <- req
+		return
+	}
+
+	h, err := s.processResponse(r)
+	if err != nil {
+		s.errCh <- err
+		return
+	}
+	log.Debugw("request headers from peer succeed", "from", s.host.ID(), "pID", stat.peerID, "amount", req.Amount)
+	// send headers to the channel, update peer stats and return peer to the queue, so it can be re-used in case
+	// if there are other requests awaiting
+	headers <- h
+	stat.updateStats(size, duration)
+	s.queue.push(stat)
+}
+
+// handeOutgoingRequest pops peer from the queue and sends a prepared request to the peer.
+func (s *session) handeOutgoingRequest(ctx context.Context, result chan []*header.ExtendedHeader) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-s.ctx.Done():
+			return
+		case req := <-s.reqCh:
+			stats := s.queue.calculateBestPeer()
+			go s.doRequest(stats, req, result)
+		}
+	}
+}
+
+// GetRangeByHeight requests headers from different peers.
+func (s *session) GetRangeByHeight(ctx context.Context, from, amount uint64) ([]*header.ExtendedHeader, error) {
+	log.Debugw("requesting headers", "from", from, "to", from+amount)
+	requests := prepareRequests(from, amount, chunkSize)
+	result := make(chan []*header.ExtendedHeader, len(requests))
+	headers := make([]*header.ExtendedHeader, 0, amount)
+	s.reqCh = make(chan *p2p_pb.ExtendedHeaderRequest, len(requests))
+
+	go s.handeOutgoingRequest(ctx, result)
+	for _, req := range requests {
+		s.reqCh <- req
+	}
+
+	for i := 0; i < cap(result); i++ {
+		select {
+		case <-s.ctx.Done():
+			return nil, errors.New("header/p2p: exchange service was stopped")
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case err := <-s.errCh:
+			return nil, err
+		case res := <-result:
+			headers = append(headers, res...)
+		}
+	}
+	sort.Slice(headers, func(i, j int) bool {
+		return headers[i].Height < headers[j].Height
+	})
+
+	return headers, nil
+}
+
+// requestHeaders sends the ExtendedHeaderRequest to a remote peer.
+func (s *session) requestHeaders(
+	ctx context.Context,
+	to peer.ID,
+	req *p2p_pb.ExtendedHeaderRequest,
+) ([]*p2p_pb.ExtendedHeaderResponse, uint64, uint64, error) {
+	now := time.Now()
+	stream, err := s.host.NewStream(ctx, to, s.protocolID)
+	if err != nil {
+		return nil, 0, 0, err
+	}
+	if err = stream.SetWriteDeadline(time.Now().Add(writeDeadline)); err != nil {
+		log.Debugf("error setting deadline: %s", err)
+	}
+	// send request
+	_, err = serde.Write(stream, req)
+	if err != nil {
+		stream.Reset() //nolint:errcheck
+		return nil, 0, 0, err
+	}
+	err = stream.CloseWrite()
+	if err != nil {
+		log.Error(err)
+	}
+	responses := make([]*p2p_pb.ExtendedHeaderResponse, 0)
```

Comment:
I would call this `headers` and call `resp` header so we know concretely what we're dealing with.